README.md
# n-send

[![Build Status](https://github.com/AlexanderMac/n-send/workflows/CI/badge.svg)](https://github.com/AlexanderMac/n-send/actions?query=workflow%3ACI)
[![Code Coverage](https://codecov.io/gh/AlexanderMac/n-send/branch/master/graph/badge.svg)](https://codecov.io/gh/AlexanderMac/n-send)
[![npm version](https://badge.fury.io/js/n-send.svg)](https://badge.fury.io/js/n-send)

### Features

- Simple configuration object
- Promise API
- Proxy support (TODO)
- Follows redirects
- Request cancelation (TODO)
- Retries on failure (TODO)
- Timeout handling
- HTTP/2 support

### Install
```bash
npm i n-send
```

### Usage
```js
const nsend = require('nsend')

try {
  const res1 = await nsend({ url: 'http://example.com' })
  // OR
  const res2 = await nsend.get('http://example.com')
} catch (err) {
  console.error(err)
}
```

### API

##### NSend Options

```js
{
  // HTTP protocol version: http/1.0, http/1.1 or http/2.0
  protocolVersion: 'http/1.1', // default

  // `baseUrl` to resolve against if the input is not absolute
  baseUrl: 'https://example.com/api',

  // `url` is the server URL that will be used for the request
  url: '/users',

  // `method` is the request method to be used when making the request
  method: 'get', // default

  // `headers` are custom headers to be sent in `name: value` format
  headers: { 'Connection': 'keep-alive' },

  // `params` are the URL parameters to be sent with the request
  // Must be a plain object
  params: {
    ts: 7567182746126
  },

  // `auth` indicates that HTTP Basic auth should be used
  // This will set an `Authorization` header, overwriting any existing
  auth: {
    username: 'smith',
    password: 'password'
  },

  // `data` is the data to be sent as the request body
  // Only applicable for request methods 'POST', 'PUT', 'PATCH'
  // Should be a `ReadStream`, `Buffer` or `String`
  data: {
    username: 'smith'
  },

  // `timeout` specifies the number of milliseconds before the request times out
  timeout: 0, // default

  // `maxContentLength` defines the max size of the http response content in bytes allowed
  maxContentLength: 10000, // default

  // `maxRedirects` defines the maximum number of redirects to follow, if set to 0, no redirects will be followed
  maxRedirects: 0 // default

  // `responseType` indicates the type of data that the server will respond with
  // options are: 'json', 'text', 'stream'
  responseType: 'json', // default

  // `responseEncoding` indicates encoding to use for decoding responses
  // Note: Ignored for `responseType` of 'stream'
  responseEncoding: 'utf8', // default
}
```

##### Request method aliases

There are aliases for all supported request methods.

```js
const nsend = require('nsend')

nsend.request(options?)
nsend.get(url, options?)
nsend.delete(url, options?)
nsend.head(url, options?)
nsend.options(url, options?)
nsend.post(url, data?, options?)
nsend.put(url, data?, options?)
nsend.patch(url, data?, options?)
```

### Licence
Licensed under the MIT license.

### Author
Alexander Mac


CHANGELOG.md
# <sub>v0.6.0</sub>
#### _Feb. 26, 2021_
  * Drop support for Node versions prior to v12.
  * Update dependencies.

# <sub>v0.5.1</sub>
#### _May. 20, 2020_
  * Fix critical issue in http2 adapter.

# <sub>v0.5.0</sub>
#### _May. 20, 2020_
  * Return `reqHeaders` object in the response for both http1 and http2 adapters.

# <sub>v0.4.0</sub>
#### _Jan. 5, 2020_
  * Add support for HTTP/2.
  * Drop support for node v8.
  * Update packages to the latest versions.
  * Refactor codebase, rewrite tests.

# <sub>v0.3.2</sub>
#### _Nov. 2, 2019_
  * Remember response redirects history.
  * Remove `content-*` headers on redirect.

# <sub>v0.3.1</sub>
#### _Nov. 2, 2019_
  * Rename response `status` to `statusCode`.

# <sub>v0.3.0</sub>
#### _Nov. 2, 2019_
  * Add support for redirects following.
  * Don't throw error when `responseType` is `json` and `response.data` can't be parsed.

# <sub>v0.2.2</sub>
#### _Sep. 17, 2019_
  * Add Type Definitions.

# <sub>v0.2.1</sub>
#### _Sep. 17, 2019_
  * Fix bug in extending nsend by method aliases.

# <sub>v0.2.0</sub>
#### _Sep. 17, 2019_
  * Add http-method aliases.

# <sub>v0.1.2</sub>
#### _Sep. 16, 2019_
  * Fix bug in transport determination.

# <sub>v0.1.1</sub>
#### _Sep. 16, 2019_
  * Replace `finally` to `then.catch` calls to support node v8.

# <sub>v0.1.0</sub>
#### _Sep. 15, 2019_
 * Release the first version.


src/consts.js
const BASE_OPTION_KEYS = {
  protocolVersion: 'protocolVersion',
  method: 'method',
  baseUrl: 'baseUrl',
  url: 'url',
  params: 'params',
  auth: 'auth',
  headers: 'headers',
  data: 'data',
  timeout: 'timeout',
  maxContentLength: 'maxContentLength',
  maxRedirects: 'maxRedirects',
  responseType: 'responseType',
  responseEncoding: 'responseEncoding'
}

const HTTP_VERSIONS = {
  http10: 'http/1.0',
  http11: 'http/1.1',
  http20: 'http/2.0'
}

const DEFAULT_OPTIONS = {
  [BASE_OPTION_KEYS.protocolVersion]: HTTP_VERSIONS.http11,
  [BASE_OPTION_KEYS.method]: 'get',
  [BASE_OPTION_KEYS.maxContentLength]: 10000,
  [BASE_OPTION_KEYS.maxRedirects]: 0,
  [BASE_OPTION_KEYS.responseType]: 'text',
  [BASE_OPTION_KEYS.responseEncoding]: 'utf8'
}

const CORE_OPTION_KEYS = [
  BASE_OPTION_KEYS.protocolVersion,
  BASE_OPTION_KEYS.method,
  BASE_OPTION_KEYS.baseUrl,
  BASE_OPTION_KEYS.url,
  BASE_OPTION_KEYS.params,
  BASE_OPTION_KEYS.auth,
  BASE_OPTION_KEYS.headers,
  BASE_OPTION_KEYS.data,
  BASE_OPTION_KEYS.timeout,
  BASE_OPTION_KEYS.maxContentLength,
  BASE_OPTION_KEYS.maxRedirects,
  BASE_OPTION_KEYS.responseType,
  BASE_OPTION_KEYS.responseEncoding
]

const ADAPTER_OPTION_KEYS = [
  BASE_OPTION_KEYS.protocolVersion,
  BASE_OPTION_KEYS.method,
  BASE_OPTION_KEYS.baseUrl,
  BASE_OPTION_KEYS.url,
  BASE_OPTION_KEYS.params,
  BASE_OPTION_KEYS.auth,
  BASE_OPTION_KEYS.headers,
  BASE_OPTION_KEYS.data,
  BASE_OPTION_KEYS.timeout,
  BASE_OPTION_KEYS.maxContentLength,
  BASE_OPTION_KEYS.responseType,
  BASE_OPTION_KEYS.responseEncoding
]

const REQUEST_OPTION_KEYS = [
  BASE_OPTION_KEYS.method,
  BASE_OPTION_KEYS.baseUrl,
  BASE_OPTION_KEYS.url,
  BASE_OPTION_KEYS.params,
  BASE_OPTION_KEYS.auth,
  BASE_OPTION_KEYS.headers,
  BASE_OPTION_KEYS.data,
  BASE_OPTION_KEYS.timeout,
  BASE_OPTION_KEYS.maxContentLength,
  BASE_OPTION_KEYS.responseType,
  BASE_OPTION_KEYS.responseEncoding
]

const RESPONSE_OPTION_KEYS = [
  BASE_OPTION_KEYS.maxContentLength,
  BASE_OPTION_KEYS.responseType,
  BASE_OPTION_KEYS.responseEncoding
]

const SAFE_METHODS = ['get', 'head', 'options', 'trace']

module.exports = {
  HTTP_VERSIONS,
  DEFAULT_OPTIONS,
  CORE_OPTION_KEYS,
  ADAPTER_OPTION_KEYS,
  REQUEST_OPTION_KEYS,
  RESPONSE_OPTION_KEYS,
  SAFE_METHODS
}
